cgrok/tree.c

```diff
@@ -0,0 +1,275 @@
+#include "tree.h"
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+// A shim to use dictlib trees instead of TC
+struct TCTREE {
+  dict *dict;
+  dict_itor *iter;
+};
+
+// Toyko cabinet accepts any arbitrary-sized data as a tree key.
+// We support this by using the first 4 bytes of the key to store the length.
+// So we have to wrap the libdict comparators with implementations that skip the first four bytes
+// of the key
+int tccmpint32(const void* k1, const void* k2) {
+  return dict_int_cmp(k1+4, k2+4);
+}
+
+int dict_var_str_cmp(const void* k1, const void* k2) {
+  return dict_str_cmp(k1+4, k2+4);
+}
+
+void tcfree(void *key, void *value) {
+  free(key);
+  free(value);
+}
+
+TCTREE *tctreenew(void) {
+  return tctreenew2(dict_var_str_cmp, 0);
+}
+
+// Make a new tree with a user-defined comparator (must be tccmpint32 or dict_var_str_cmp),
+// because of how we pack the keys
+TCTREE *tctreenew2(dict_compare_func cp, void *cmpop) {
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  if (!tree) {
+    return NULL;
```

Comment:
No, grok doesn't actually check the return value at all. This was part of the larger discussion of how we should handle malloc failing.